keps/sig-network/20191205-kube-proxy-IP-node-binding.md

```diff
@@ -0,0 +1,70 @@
+---
+title: Give the ability to the LoadBalancer provisioner to disable the LB IP binding
+authors:
+  - "@Sh4d1"
+owning-sig: sig-network
+participating-sigs:
+  - sig-cloudprovider
+reviewers:
+  - "@thockin"
+  - "@cheftako"
+  - TBD
+approvers:
+  - "@thockin"
+  - TBD
+editor: TBD
+creation-date: 2019-12-05
+last-updated: 2019-12-05
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Give the ability to the LoadBalancer provisioner to disable the LB IP binding
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+<!-- /toc -->
+
+## Summary
+
+The different kube proxy implentations (at least ipvs and iptables) are today sort of binding the External IPs of the LoadBalancer Service to each node. (iptables is creating some rules to redirect directly the packets to the service and ipvs is binding the IP to one interface on the node). This feature is exisiting because:
+
+1. To redirect pod traffic that is intended for LoadBalancer IP directly to the service, bypassing the loadbalancer for path optimization.
+2. Some LBs are sending packet with the destination IP equal to the LB IP, so the IP needs to be routable directly to the right service (otherwise looping back and forth)
+3. (TBD)
+
+This enhancement proposes a way to make this features configurable, giving the cloud provider, or the user, a way to disable this behaviour of kube proxy.
+
+## Motivation
+
+There are numerous problems with the current behaviour:
+
+1. Some cloud provider (DigitalOcean, Scaleway, ...) are using the LB's external IP as egress IP. This is a probem in the ipvs mode of kube proxy since the IP is bouded to an interface and healtchecks from the LB are never coming back.
```

Comment:
I may be misunderstanding this statement, or it might need updated with regard to the DigitalOcean portion.  I'm an engineer with DigitalOcean so, I'll try to clarify how things work on our end.  

DigitalOcean load-balancers aren't used for egress. For traffic egressing from the cluster, each node has its own public IP.  Traffic from pods on that node (and healthchecks from the kubelet) are SNAT'ed and egress with that public IP.  Traffic from the load-balancer to the cluster (health checks and forwarded connections) have a source-IP on the private network interface (10.x.x.x) and the destination IP is the node ip.  The reverse flow travels back to the LB (no DSR).  Because the cluster (and it's pods) see the source IP as the private address of the load-balancer, we offer ProxyProtocol support so customer workloads (ie. ingress controller) can determine the real source IP.

The second point however is definitely relevant to DigitalOcean.
>Some cloud provider are also having features at the LB level (TLS termination, PROXY protocol, ...) and bypassing the LB means missing these features when the packet arrives to the service (leading to protocols errors).

Reply:
@jcodybaker yeah formulation was not exactly right, by egress IP I meant the IP from the LB to the pod. But fair enough for the private IP, I'll update this point tomorrow :)

Reply:
Gotcha that makes sense.  The private IP of the load-balancer isn't referenced in the svc.status.loadbalancer and therefore should be unimpacted by these kube-proxy rules.  The statement may still apply to scaleway however.

Thanks for clarifying.  Please feel free to reach out to me here or cody.baker on Kubernetes slack if DO can be of any assistance with this issue.

Reply:
I think using LB IP for egress is independent from saying all access to the LB IP (irrespective of from within the cluster or outside the cluster) needs to go to the LB IP and back and not locally route. Isn't that true?

Reply:
Again, I really need to change this phrasing. Not really talking about LB IP for egress traffic but when the source IP of the traffic from LB to cluster is the same as the external IP published in the service